### PR TITLE
Make Slack notifications pretty

### PR DESCRIPTION
**Pull request by nickoferrall**
PR to resolve issue #4911. 

It also resolves an undocumented bug where timer adjustments send new notifications rather than update the previous one. I've commented inline about this. 

A few additional notes:

- The issue mentioned using Slack's default green `#007a5a` but with Slack's blocks, you choose whether it's a primary button, regular or danger button rather than specify the colour. On a dark mode theme, we get the darker button and in a light mode, it's a lighter button
- I haven't come across a good, free tool for creating short links so I've included the regular link without utm params which is still shortish
- I've added a dummy interactivity request URL to the [Slack app](https://api.slack.com/apps/A01EGHDU7PZ/interactive-messages?). Otherwise, a warning sign shows after clicking the button. See Slack issue [here](https://github.com/CircleCI-Public/slack-orb/issues/188#issuecomment-708988009)
- I noticed that the comment count is currently `undefined` when ending poker meetings. I've created an issue to fix this [here](https://github.com/ParabolInc/parabol/issues/4983). When it's fixed, we can add comment count to the poker notifications

### To test

- Go to the `.env` file and change the `SLACK_CLIENT_ID` & `SLACK_CLIENT_SECRET` to the client ID and secret found in the `Parabol-staging` bot [here](https://api.slack.com/apps/A01EGHDU7PZ)
- Kill and restart the server with `yarn dev`
- You may need to install the Parabol-staging Slackbot in Slack
- Integrate Slack in Parabol (remove it first if it's already integrated),
- See the new notifications styles when you create, set a timer & end a meeting
- If the notifications are sent to `@Parabol-staging` the previous notification should be updated when you adjust the timer rather than sending a new notification

### CODE REVIEW

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (nickoferrall) [inline packages/server/graphql/mutations/helpers/notifySlack.ts:112]: To keep the link shortish, I've removed the utm params. Hopefully, this isn't too problematic for our analytics?

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (nickoferrall) [inline packages/server/graphql/mutations/helpers/notifySlack.ts:136]: Removed the check for the `Parabol` bot name as `latest` only exists in the DM and we handle these checks server side. Small advantages being it's slightly cleaner and we don't need to change the name to `Parabol-staging` in development

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (nickoferrall) [inline packages/server/graphql/mutations/helpers/notifySlack.ts:215]: I noticed that when changing the timer, a new notification is created rather than the previous one being edited. The reason is in `addSlackAuth`, we're using the Slack user id as the `channelId` so that we can send messages from the `@Parabol`. However, to be able to update a message, we need the DM `channelId`. 

Requesting user info on the bot or bot info doesn't return this so I've grabbed all the IMs and found the IM `channelId` with the Slackbot's user id (see [here](https://github.com/ParabolInc/parabol/blob/a78aa53259e1c303e5f0cd82ec393bf46d346219/packages/server/graphql/mutations/addSlackAuth.ts#L119)). 

To be able to read all the IM channels, I think we should need the `im:read` [scope](https://api.slack.com/scopes/im:read). While this works in development without the scope, I suspect this is the same bug as the private channels one where the dev & prod bots require different scopes.

I haven't adjusted the scopes in this PR. If the `im:read` scope is required, it will simply post new notifications when the timer is adjusted rather than update the old one.

Instead, I'll adjust the scope in [this PR](https://github.com/ParabolInc/parabol/pull/4964) so that we don't need to wait until the Slack app is approved to merge this PR.

**Comment by enriquesanchez** [comment]
@nickoferrall I was not able to get this to work on my end, but I'm seeing the notifications you created on the `test-slack-bot` channel. It looks much better 🎉

>The issue mentioned using Slack's default green #007a5a but with Slack's blocks, you choose whether it's a primary button, regular or danger button rather than specify the colour. On a dark mode theme, we get the darker button and in a light mode, it's a lighter button

This is totally OK. Thanks for looking into it.

>I haven't come across a good, free tool for creating short links so I've included the regular link without utm params which is still shortish

We already use our own shortlinks, such as `https://prbl.in/J12QXX12T4IZ` when you click on "Invite to team". Can we use that for the notifications as well?

**Comment by enriquesanchez** [comment]
One more thing:

Screen Shot 2021-05-21 at 3 26 12 p m


Can we move the `Meeting` value a bit more to the left, so it's closer to `Team`? Something like this:

Screen Shot 2021-05-21 at 3 26 12 p m

**Comment by nickoferrall** [comment]
> We already use our own shortlinks when you click on "Invite to team"

Ah thanks! I checked this in dev where we show the regular link but I see in production we have the short link. I've pushed an update with this

> Can we move the Meeting value a bit more to the left, so it's closer to Team?

Slack blocks aren't really built for customisation so this is tricky. You define fields that you want to use and they apply their own [styling](https://app.slack.com/block-kit-builder/T08FL6336#%7B%22blocks%22:%5B%7B%22type%22:%22section%22,%22text%22:%7B%22type%22:%22mrkdwn%22,%22text%22:%22You%20have%20a%20new%20request:%5Cn*%3CfakeLink.toEmployeeProfile.com%7CFred%20Enriquez%20-%20New%20device%20request%3E*%22%7D%7D,%7B%22type%22:%22section%22,%22fields%22:%5B%7B%22type%22:%22mrkdwn%22,%22text%22:%22*Type:*%5CnComputer%20(laptop)%22%7D,%7B%22type%22:%22mrkdwn%22,%22text%22:%22*When:*%5CnSubmitted%20Aut%2010%22%7D,%7B%22type%22:%22mrkdwn%22,%22text%22:%22*Last%20Update:*%5CnMar%2010,%202015%20(3%20years,%205%20months)%22%7D,%7B%22type%22:%22mrkdwn%22,%22text%22:%22*Reason:*%5CnAll%20vowel%20keys%20aren't%20working.%22%7D,%7B%22type%22:%22mrkdwn%22,%22text%22:%22*Specs:*%5Cn%5C%22Cheetah%20Pro%2015%5C%22%20-%20Fast,%20really%20fast%5C%22%22%7D%5D%7D,%7B%22type%22:%22actions%22,%22elements%22:%5B%7B%22type%22:%22button%22,%22text%22:%7B%22type%22:%22plain_text%22,%22emoji%22:true,%22text%22:%22Approve%22%7D,%22style%22:%22primary%22,%22value%22:%22click_me_123%22%7D,%7B%22type%22:%22button%22,%22text%22:%7B%22type%22:%22plain_text%22,%22emoji%22:true,%22text%22:%22Deny%22%7D,%22style%22:%22danger%22,%22value%22:%22click_me_123%22%7D%5D%7D%5D%7D). Rather than show two fields, i.e. Team and Meeting, we could show one field with Team and Meeting separated by a bunch of spaces but this would be kinda hacky and difficult to maintain on desktop and mobile. Unless you have a lot of tension with this one, I suggest we stick with Slack's format

**Comment by enriquesanchez** [comment]
@nickoferrall Thank you!

>we could show one field with Team and Meeting separated by a bunch of spaces but this would be kinda hacky and difficult to maintain on desktop and mobile. Unless you have a lot of tension with this one, I suggest we stick with Slack's format

Agreed, let's go with the defaults Slack allows. Thank you for looking into it.

**Review by enriquesanchez — APPROVED** [review]

File: packages/server/graphql/mutations/endRetrospective.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/endRetrospective.ts:124]: this isn't safe because we need to wait for `removeEmptyTasks` probably worth reverting the change & adding a comment as to why it's done in serial

File: packages/client/utils/SlackManager.ts
Comment (mattkrick) [inline packages/client/utils/SlackManager.ts:227]: blocks must be a string (you pass it in correctly, but the type can't be an array)

File: packages/server/graphql/mutations/helpers/makeSlackBlocks.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/makeSlackBlocks.ts:25]: does this have to be a UID? if so, it won't be unique enough because this could execute twice in the same millisecond.
i'd just use `generateUID`

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/notifySlack.ts:209]: alternatively, you could make `updateMessage` take the same type as postMessage so they have the same call signature

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (mattkrick) [reply packages/server/graphql/mutations/helpers/notifySlack.ts:214]: just outta curiosity, have you tried querying that one specific channel with something like conversations.info? it'd be absolutely amazing to avoid an extra scope if possible

**Review by mattkrick — CHANGES_REQUESTED** [review]
couple light changes, shouldn't take long! if this misses the ship window we can always do a 2nd release just for this

**Comment by nickoferrall** [comment]
> just outta curiosity, have you tried querying that one specific channel with something like conversations.info? it'd be absolutely amazing to avoid an extra scope if possible

Yup, I've tried [conversations.info](https://api.slack.com/methods/conversations.info) (just tested it again to be sure) but it requires a `channelId` which we don't have for the Slackbot. The incoming webhook response doesn't provide the Slackbot `channelId` and nor does [auth.test](https://api.slack.com/methods/auth.test), [bots.info](https://api.slack.com/methods/bots.info) or [users.info](https://api.slack.com/methods/users.info). It looks like we have to search through the IMs to get the Slackbot `channelId`. 

Alternatively, if we want to avoid requesting the `im:read` scope, we could scrap the update message logic. The update message logic happens when a user updates the timer and the default Slack channel is the `@Parabol` bot. If it's a default Slack channel is a group channel, we send a new message as we don't get the timestamp of the latest message. Sparing a user a notification from `@Parabol` is nice but the value add is quite small IMO

File: packages/client/utils/SlackManager.ts
Comment (nickoferrall) [reply packages/client/utils/SlackManager.ts:227]: Thanks! I've moved the stringify to `updateMessage` so that we can have the same call signature here and keep the array type

File: packages/server/graphql/mutations/endRetrospective.ts
Comment (nickoferrall) [reply packages/server/graphql/mutations/endRetrospective.ts]: Ahh yes! I guess this is the same for [check-ins too](https://github.com/ParabolInc/parabol/blob/8d3826bfaedeb8d17b08afbd9bda56866eace4f6/packages/server/graphql/mutations/endCheckIn.ts#L231) where we're removing empty tasks and finishing the meeting in the same `Promise.all`. I've moved `finishCheckInMeeting` too and added comments in both files 👍

File: packages/client/utils/SlackManager.ts
Comment (mattkrick) [inline packages/client/utils/SlackManager.ts:228]: only stringify if array, otherwise we'll get double quotes

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/notifySlack.ts:116]: is `https:/prbl.in/${meetingId}` valid?

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/notifySlack.ts:123]: for discriminated unions like this it's a little cleaner to switch on the type
you can use a switch or a condition, both give full type safety as long as you use object access & that way if reflectionCount goes away, we don't have to change the logic:

```ts
  switch (meeting.meetingType) {
    case 'retrospective':
      
  }
  if (meeting.meetingType === 'retrospective') {

  }

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/notifySlack.ts:169]: discuss only exists on retros, what about check-ins?

i see a lot of conditional logic based on meeting type. might be worthwhile branching this logic based on meetingType.

e.g. `const blocks = makeBlocksForMeeting()`
makeBlocksForMeeting = () => {
  switch (meeting.meetingType) {
    case 'retro': 
    case 'action':
      return foo
  }
}
```

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/notifySlack.ts:261]: https:/prbl.in/${meetingId} i dont think this is valid

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (nickoferrall) [reply packages/server/graphql/mutations/helpers/notifySlack.ts:116]: The url isn't valid but it's just a hyperlink. When the user clicks on it, they'll be redirected to the regular longer link with UTM params. The url could be anything but I kept it in the same format as our team invite link. 

If we want to use a valid short url using our existing format of `https:/prbl.in/${token}`, we'd run into a security issue as the token works as an invite to a team and a user could send the Slack notification to a public channel

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (nickoferrall) [reply packages/server/graphql/mutations/helpers/notifySlack.ts]: I tried to use the meeting type initially but it means we have to cast the type (which we do in [joinMeeting](https://github.com/ParabolInc/parabol/blob/4ce15d9ba712b4004d6361be3c5daee5714e0ad9/packages/server/graphql/mutations/joinMeeting.ts#L29)). We don't need to cast the type when using  `reflectionCount` but I definitely agree that using a field that could be removed at some point isn't ideal

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (nickoferrall) [reply packages/server/graphql/mutations/helpers/notifySlack.ts:246]: I responded to this in the PR thread but adding it here incase you didn't see it: 

Yup, I've tried [conversations.info](https://api.slack.com/methods/conversations.info) (just tested it again to be sure) but it requires a `channelId` which we don't have for the Slackbot. The incoming webhook response doesn't provide the Slackbot `channelId` and nor does [auth.test](https://api.slack.com/methods/auth.test), [bots.info](https://api.slack.com/methods/bots.info) or [users.info](https://api.slack.com/methods/users.info). It looks like we have to search through the IMs to get the Slackbot `channelId`. 

Alternatively, if we want to avoid requesting the `im:read` scope, we could scrap the update message logic. The update message logic happens when a user updates the timer and the default Slack channel is the `@Parabol` bot. If it's a default Slack channel is a group channel, we send a new message as we don't get the timestamp of the latest message. Sparing a user a notification from `@Parabol` is nice but the value add is quite small IMO

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (nickoferrall) [reply packages/server/graphql/mutations/helpers/notifySlack.ts]: True, it looks like most of the conditional logic is in the end meeting notification. I've created a couple of functions to make it a bit more readable

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (mattkrick) [reply packages/server/graphql/mutations/helpers/notifySlack.ts:116]: feels a little weird that the link label is an invalid url. could we instead write something like: [Take me there](https://www.google.com)?

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (mattkrick) [reply packages/server/graphql/mutations/helpers/notifySlack.ts]: well then i made those classes poorly.
For each meeting class (e.g. MeetingRetrospective) could you add the meeting type as a string const?
for example

```ts
export default class MeetingRetrospective extends Meeting {
  meetingType!: 'retrospective'
}
```

after you've done that for all 3 you'll be able to discriminate on the value of `meetingType`

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (nickoferrall) [reply packages/server/graphql/mutations/helpers/notifySlack.ts:116]: For sure! I thought this was the approach your were suggesting [here](https://parabol.slack.com/archives/D017D764FDY/p1621887268010200).

The "Join meeting" button already takes us to the link so perhaps we just drop the "Link" section?

notification

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (nickoferrall) [reply packages/server/graphql/mutations/helpers/notifySlack.ts]: Cool thanks - I've pushed that update and it's working nicely

File: packages/server/graphql/mutations/helpers/notifySlack.ts
Comment (enriquesanchez) [reply packages/server/graphql/mutations/helpers/notifySlack.ts:116]: >The "Join meeting" button already takes us to the link so perhaps we just drop the "Link" section?

@nickoferrall I added the link section on purpose because I think having the link there could be handy for cases when you need to share the link with someone, for example on a text message, or an email. You can't copy the link from the "Joing meeting" button, unless you join the meeting first and then copy the link from within Parabol, which I thought would be a bit annoying if joining the meeting is not what you want to do at the moment.

**Comment by mattkrick** [comment]
LGTM! so stoked to see these new notifications in our channel!

**Pull request merged**
